### ISSUE TITLE
Improve author archive?

### ISSUE THREAD

**Issue by briggySmalls** [jerome-toole/skindeep#2]
It is not currently possible to view all posts for a given (custom) author.
Whilst the `<root>/author/<author_name>` endpoint is live, it expects parameter `<author_name>` to be a wordpress username (stock behaviour).
As a result all posts appear under `<root>/author/skindeep`.
Given posts are credited to the custom author field (not a user account), should the author archive behaviour be reconfigured to be parameterised by the custom author field?

**Comment by jerome-toole** [comment]
hmmmm. the problem we have here is that the author field isn't a taxonomy, it's just a custom field so it's not set up to be indexable.

We'd need to set up an author taxonomy (and I guess an illustrator taxonomy too) in order to use all the good features of a category, tag etc.

To do this, we'd add something like this to the `functions.php` file.

```php

add_action( 'init', __NAMESPACE__ . '\\create_author_taxonomy' );

function create_author_taxonomy() {
  $labels = array(
    'name'                           => 'Authors',
    'singular_name'                  => 'Author',
    'search_items'                   => 'Search Authors',
    'all_items'                      => 'All Authors',
    'edit_item'                      => 'Edit Author',
    'update_item'                    => 'Update Author',
    'add_new_item'                   => 'Add New Author',
    'new_item_name'                  => 'New Author Name',
    'menu_name'                      => 'Author',
    'view_item'                      => 'View Author',
    'popular_items'                  => 'Popular Author',
    'separate_items_with_commas'     => 'Separate authors with commas',
    'add_or_remove_items'            => 'Add or remove authors',
    'choose_from_most_used'          => 'Choose from the most used authors',
    'not_found'                      => 'No authors found'
  );
  register_taxonomy(
    'author',
    'post',
    array(
      'label' => __( 'Author' ),
      'hierarchical' => false,
      'labels' => $labels,
      'public' => true,
      'show_in_nav_menus' => false,
      'show_tagcloud' => false,
      'show_admin_column' => true
    )
  );
}

```

You'll then have a box like this on all posts:

![image](https://cloud.githubusercontent.com/assets/5527867/20732825/60346444-b688-11e6-85f0-759c57e3e663.png)

We can then add an image, description or other meta data to each taxonomy term (author) in future. Pretty sure that ACF can handle this.

**Comment by briggySmalls** [comment]
Interesting. That sounds much better than the approach I was looking at.

I found that adding the following to the functions.php would change the query executed when accessing the `<root>/author/<author_name>` endpoint to searching for posts that have a meta author with `<author_name>`.

```php
/**
 * Modify any requests for posts for a particular author and use our modified
 * author instead
 *
 * @param      WP_Query  $query  The original query constructed by wordpress
 *
 * @return     WP_Query  The modified query that fetches posts by modified author
 */
function mod_author_request( $query ) {
    if ( is_admin() || ! $query->is_main_query() )
        return;
    if ( is_author() ) {
        // Use modified author found in postmeta
        $query->set('meta_key', 'author');
        $query->set('meta_value', str_replace("-", " ", $query->get('author_name')));
        $query->set('author_name', 'skindeep');
        return $query;
    }
}
add_action( 'pre_get_posts', 'mod_author_request', 1 );
```

While this works, it is obviously way more hacky. The string replacement is particularly nasty (I just checked and searching for `<root>/author/Andrej-Mircev` works when `<root>/author/Andrej-Mirčev` doesnt...which is both fortuitous and odd...direct SQL queries work for both...).

So the attractions are:
1. It works straight away
2. It requires no user maintenance of a taxonomy.

But the disadvantages are:
1. It is ugly and short-termist. If the requirement for an author image/description etc developed it would not be appropriate.

**Comment by jerome-toole** [comment]
What do you mean by 'requires no user maintenance of a taxonomy'? For the user, it would be the same as writing the author is now but the added benefit is you would get autocomplete for authors written in the past.

The only disadvantage is that someone has to go through and copy the author names on all the articles. Someone will have to go through and add the illustrators anyway so maybe this is something they can do at the same time.

**Comment by briggySmalls** [comment]
Ah apologies. You're completely right.
I didn't realise that the taxonomy could be added to just by saving on a post. This is like how tags are created/assigned? This would be the preferred way to do this?
ACF off-the-shelf taxonomy fields seem to be of the 'select from drop down' or 'checkbox in list' form, etc. It looks like you need to add new legitimate values for the taxonomy elsewhere before they appear in the drop down. That's sort of what I meant by maintenance. With those you can't simply create a new author at the same time as making a post...I think.
I'll look into using ACF (or otherwise) to assign, and if necessary create, an author/illustrator on a post in the same style as tags are.
Cheers!

**Comment by briggySmalls** [comment]
Slightly related. By having the _%category_name%/%post_name%_ permalinks we expose more than just the author endpoint. We also expose the 'donor' endpoint, as registered by __Seamless Donations__.
This is not great for donor privacy. When fixing up the taxonomies, these endpoints might need reviewing?

**Comment by jerome-toole** [comment]
Yup the authors would be just like tags. We wouldn't use ACF for that. You just register a new taxonomy (like tags) which happens to be called Authors and then it functions just like tags.

We would then use ACF to add a field to the taxonomy page (accessible in Posts > Authors) which allows an image etc. to be added.

Does this make sense now?

**Comment by jerome-toole** [comment]
@briggySmalls have you tried the above `register_taxonomy` code in functions.php?

If I'm not mistaken, as soon as you make that, you'll be able to add author names and then there should automatically be archives... Let me know if not

**Comment by briggySmalls** [comment]
Hi @jerome-toole.

Sorry I have not tried this out. I'll get on this next (along with #23)

**Comment by briggySmalls** [comment]
This works. But using 'author' we get clashes with the standard 'author' attribute (numerical ID) in some cases. E.g. if I select the 'author' taxonomy value (in this case 'Skin') of a post in the *wp-admin/edit.php* view, by selecting it from the new column, the page is loaded with a parameter *wp-admin/edit.php?author=skin*.
It would appear no filtering gets performed as it is assumed that we are targeting the author ID and the text 'skin' does not parse to a number.

If I change the taxonomy name to 'tauthor' the page is loaded with a parameter *wp-admin/edit.php?tauthor=skindeep*. The filtering then works as expected.

<img width="1322" alt="screen shot 2017-03-08 at 21 04 30" src="https://cloud.githubusercontent.com/assets/11054133/23723930/e35e5098-0442-11e7-90e6-70efc107ef71.png">

I suppose this doesn't matter so much for the public side of the site as we would modify our permalinks structure to support author archives? Something like: *__<base>__/articles/author/__<tax_author_name>__*?
